eviction: regression test + distinguish layer write from map insert

problame opened the pull request:
This patch adds a regression test for the threshold-based layer eviction.
The test asserts the basic invariant that, if left alone, the residence statuses will stabilize, with some layers resident and some layers evicted.
Thereby,  we cover both the aspect of last-access-time-threshold-based eviction, and the "imitate access" hacks that we put in recently.

The aggressive  `period` and `threshold` values revealed a subtle bug which is also fixed in this patch.
The symptom was that, without the Rust changes of this patch, there would be occasional test failures due to `WARN... unexpectedly downloading` log messages.
These log messages were caused by the "imitate access" calls of the eviction task.
But, the whole point of the "imitate access" hack was to prevent eviction of the layers that we access there.
After some digging, I found the root cause, which is the following race condition:

1. Compact: Write out an L1 layer from several L0 layers. This records residence event `LayerCreate` with the current timestamp.
2. Eviction: imitate access logical size calculation. This accesses the L0 layers because the L1 layer is not yet in the layer map.
3. Compact: Grab layer map lock, add the new L1 to layer map and remove the L0s, release layer map lock.
4. Eviction: observes the new L1 layer whose only activity timestamp is the `LayerCreate` event.

The L1 layer had no chance of being accessed until after (3).
So, if enough time passes between (1) and (3), then (4) will observe a layer with `now-last_activity > threshold` and evict it

The fix is to require the first `record_residence_event` to happen while we already hold the layer map lock.
The API requires a ref to a `BatchedUpdates` as a witness that we are inside a layer map lock.
That is not fool-proof, e.g., new call sites for `insert_historic` could just completely forget to record the residence event.
It would be nice to prevent this at the type level.
In the meantime, we have a rate-limited log messages to warn us, if such an implementation error sneaks in in the future.

fixes https://github.com/neondatabase/neon/issues/3593
fixes https://github.com/neondatabase/neon/issues/3942

problame reviewed (commented):
Should I duplicate the explainer of the race condition somewhere in the code, as a comment? If so, where?

koivunej commented:
> `WARN... unexpectedly downloading`

I think we saw this in logs but probably forgot to add a note. But good that you re-hit that.

koivunej commented:
> So, we have to do it at the call sites of `insert_historic`.

Makes me think the #3775 fix or wrapping is more needed than not.

koivunej commented on test_runner/regress/test_threshold_based_eviction.py:
```diff
@@ -0,0 +1,183 @@
+
+    log.info("len(map_info_changes)=%s", len(map_info_changes))
+
+    # TODO: can we be more precise here? E.g., require we're stable _within_ X*threshold,
+    # instead of what we do here, i.e., stable _for at least_ X*threshold toward the end of the observation window
```
Feels like this could be unnecessary flakyness issue with multiple test running at the same time. I don't have a better idea.

problame replied:
Yeah. From what I've seen in the test i've run locally, it's always stable within one `threshold` iteration.
Would be nice to have a `generation number` for the layer map, and expose it in the layer map info.
That way, we can just look at the generation number instead of comparing the layer maps.

Perhaps the wrapper in #3775 could add this, to keep the main layer map code clean.

koivunej replied:
Agreed, version or generation would be good. Also related to gc/compaction + upload waiting...

koivunej replied:
Resolving after linking your comment on #3775.

koivunej commented on test_runner/regress/test_threshold_based_eviction.py:
```diff
@@ -0,0 +1,183 @@
+            map_info_changes.append(current)
+            log.info("change in layer map\n before: %s\n after: %s", last, current)
+        else:
+            stable_for = current[0] - last[0]
+            log.info("residencies stable for %s", stable_for)
+            if stable_for > consider_stable_when_no_change_for_seconds:
+                break
```
There's a chance of "waiting for nothing" at seconds 6, 7, 8 (or 5, 6, 7) where we cannot go `stable_for > consider_stable_when_no_change_for_seconds`. But I cannot see how could you avoid that elegantly.

I am a rather confused why not `>=` but `>`, but it's consistent so not really an issue.

problame replied:
> There's a chance of "waiting for nothing" at seconds 6, 7, 8 (or 5, 6, 7) where we cannot go stable_for > consider_stable_when_no_change_for_seconds. But I cannot see how could you avoid that elegantly.

Hm, not sure I understand what you mean by "waiting for nothing".

koivunej replied:
I meant that on those rounds if there's change, you can no longer succeed in the check, because you will not be able to accumulate 4 successive rounds. But, no elegant solution I could think of.

As in:

1. change
2. change
3. change
4. change
5. change
6. same as 5
7. same as 5
8. same as 5

If there's a change on 5th, then the others are needless.

koivunej replied:
Idle chatter, as we cannot see a better alternative. Resolving.

koivunej reviewed (approved):
Mostly trusting the PR desc, code matching it and the test passing, looking good. Comments mostly questions.

problame commented:
> Makes me think the https://github.com/neondatabase/neon/pull/3775 fix or wrapping is more needed than not.

Yeah, had that thought as well.

koivunej commented on pageserver/src/tenant/timeline.rs:
```diff
@@ -4092,7 +4099,14 @@ impl Timeline {
+                Err(error) => {
+                    warn!(%error, layer=%l.filename().file_name(), "latest activity not available, using UNIX epoch as fallback");
+                    // If we're under disk pressure, we should not hide this layer's existence.
+                    SystemTime::UNIX_EPOCH
```
> How about returning the LayerCreateFileWritten timestamp in the Err(), and using that?
That way, a layer that's truly brand-new will be at a top.

That works just as well, might be a more accurate if one needs to hunt down timestamps.

problame replied:
Obsolete.

problame commented:
Changes:
1. Implemented an alternative approach that puts more responsibility to the layer-creating code. Pushes some checks into type system by requiring a witness. See commit [feat: alternative approach that takes a witness for layer map lock held](https://github.com/neondatabase/neon/pull/4005/commits/72fad664e3fad94d73ae95ccb1baac1c9e73fb89).
1. Rebased (insert_historic can now fail, had conflicts, CI wouldn't run anymore on this branch)

@koivunej WDYT, is the new approach better?

koivunej commented on pageserver/src/tenant/storage_layer.rs:
```diff
@@ -158,41 +159,82 @@ impl LayerAccessStatFullDetails {
+    /// using [`record_residence_event`] before calling `latest_activity`.
+    /// If they don't, [`latest_activity`] will return `None`.
+    pub(crate) fn empty_will_record_residence_event_later() -> Self {
+        LayerAccessStats(Mutex::new(LayerAccessStatsLocked::default()))
```
Just FYI, this could be simplified as Mutex's and RefCells and pretty much all types with Default impl also use the inner types Default::default().

Most terse I'd be fine here:

```
LayerAccessStats(Default::default())
```

But in general we hope to avoid fully generic, so

```suggestion
        LayerAccessStats(Mutex::default())
```

Would work just as well. But the current is fine as well.

problame replied:
TBH I never got why we don't use `Default::default()`

koivunej commented on test_runner/regress/test_metric_collection.py:
```diff
@@ -24,13 +24,6 @@
 from werkzeug.wrappers.request import Request
 from werkzeug.wrappers.response import Response
 
-
-@pytest.fixture(scope="session")
-def httpserver_listen_address(port_distributor: PortDistributor):
-    port = port_distributor.get_port()
-    return ("localhost", port)
-
-
```
I wonder if this went wrong way in rebase? @LizardWizzard probably changed this.

koivunej replied:
`main` has these lines: https://github.com/neondatabase/neon/blob/b6c7c3290f795d918f726de2df0015bb4a3cd260/test_runner/regress/test_metric_collection.py#L28-L31

koivunej replied:
No, it's not a recent change.

problame replied:
No, I moved this out of `test_metric_collection.py` into `neon_fixtures.py`. He didn't change this.

koivunej replied:
Uh ok ... Unsure if this is related? I missed how.

koivunej replied:
Reason was that the added test wants to have an metric endpoint to cause metric task to query synthetic sizes in it's normal fashion.

koivunej reviewed (approved):
Still on the approval stance. I don't see a huge difference, but the `witness` certainly does guide towards the wanted implementation. At the same time it cannot guide towards *not-leaving the recording out*, which doesn't really fit in our current types. Would require some two-step builder.

But it's not like we are trying to make a fool-proof library out of this, and the only downside is disk based eviction giving out a warning.

koivunej commented on pageserver/src/tenant/timeline.rs:
```diff
@@ -4107,7 +4131,15 @@ impl Timeline {
+                None => {
+                    // If we're under disk pressure, we should not hide this layer's existence.
+                    // Log a warning and
+                    warn!(layer=%l.filename().file_name(), "latest activity not available, did we forget to add a residence event?");
```
Re: `%l.filename().file_name()`:

I think we should try to refactor a `Display` impl to Layer in addition to the verbose `Debug` and move towards using just these. I am pretty sure we discussed this recently and I may have gone with `layer_id` then. Or maybe it was for the absolute path logging issue, to use layerid instead of file name.

problame replied:
Agree, separate issue though. Can you create one?

koivunej replied:
Created https://github.com/neondatabase/neon/issues/4145

problame commented:
@koivunej in reaction to your comment on the log message, I pushed
[feat: rate-limit logging + centralize in the latest_activity method f](https://github.com/neondatabase/neon/pull/4005/commits/f480844ccfc7c07b05177782c0563e97a6b76564) , I think this is the max amount of yak-shaving I'm able to tolerate ;)

koivunej commented on pageserver/src/tenant/timeline.rs:
```diff
@@ -4131,15 +4131,8 @@ impl Timeline {
-                }
-            };
+            // Use UNIX_EPOCH as a fallback timestamp in case we have a bug and forget to record a residence event for the layer.
+            let last_activity_ts = l.access_stats().latest_activity(|| SystemTime::UNIX_EPOCH);
```
Keeping the warning here would have the benefit of warning just once (and maybe total) per eviction.

For timelines eviction_task yeah this is a bit more unfortunate, there we definitely don't want to see a lot of warnings every round.

koivunej commented on pageserver/src/tenant/timeline.rs:
```diff
@@ -4115,10 +4132,11 @@ impl Timeline {
+                Some(ts) => ts,
+                None => {
                     // If we're under disk pressure, we should not hide this layer's existence.
+                    // Log a warning and
```
Stray comment.

koivunej replied:
I guess this got rewritten because no resolution and message??

koivunej commented on pageserver/src/tenant/storage_layer.rs:
```diff
@@ -255,24 +300,44 @@ impl LayerAccessStats {
+                    let occurences = guard.0;
+                    guard.1.call(move || {
+                        // Spawn a new task to have a new tracing root. It's ok, it only happens every 10 secs.
+                        tokio::spawn(async move {
+                            warn!(occurences, "latest_activity not available, this is an implementation bug, using fallback value");
+                        });
```
What is the point of spawning here compared to just doing it within the closure?

koivunej replied:
This question still holds.

koivunej replied:
```suggestion
                        warn!(parent: None, occurences, "latest_activity not available, this is an implementation bug, using fallback value");
```

This should do the trick, if it really was to be detached, but I don't understand what is the rationale? We can query logs with regex.

problame replied:
Sorry, I thought I had replied here.

Yeah, the point of the spawn was to log in global scope.

I always forget about the `parent: None`, will use that.

koivunej reviewed (changes requested):
I preferred the non-last minute version; having to require the ratelimited logging is a problem in itself caused by the fact how we currently don't always have a last access timestamp for what I understand, api design reasons. It wasn't a problem in the past so I already forgot why is it a problem now.

I preferred the pre-ratelimited version at `f480844^` which I already approved; the logging could be done at the callsites (for disk usage based eviction) and maybe for eviction task we could handle it in some summary if at all.

Looking at this it will take a lot longer to understand what is the case where it remains `None`, I think I proposed `SystemTime::now` for that which I can now find from the other.

problame commented:
> I preferred the non-last minute version; having to require the ratelimited logging is a problem in itself caused by the fact how we currently don't always have a last access timestamp for what I understand, api design reasons. It wasn't a problem in the past so I already forgot why is it a problem now.

We never expect to hit the log message because we're careful to add the `record_residence_event` calls in all the right places.
However, if we forget that in a branch not covered by tests (e.g., in a future refactoring), then the risk is that we're hitting it for many layers, thereby flooding the ERROR log.

> the logging could be done at the callsites (for disk usage based eviction) and maybe for eviction task we could handle it in some summary if at all.

For disk-usage-based eviction, rate-limiting to once per iteration is fine.

For the per-timeline eviction task, rate-limiting to once per ieration still leaves `$num_timelines/$period` ERROR logs per pageserver. E.g., `2000msg/10min` = 3msg/sec`.

IMO that's too much, hence the global rate limit.
It's an implementation issue after all.
It affects all call sites similarly.
Hence IMO it's nonsense to log at the callsites, esp. if one of them (per-timeline eviction task) will require rate limiting.

The ugliness of the rate limiting is constrainted to the impl, where we have a "nice" comment explaining the situation.

> Looking at this it will take a lot longer to understand what is the case where it remains None, I think I proposed SystemTime::now for that which I can now find from the other.

The `warn` message will still print every 10 seconds. It really takes no genius to grep for the message in the codebase, then do git blame / read surrounding code, which is well-commented. So, I think it's easy to understand what's going on.

---

Does the above convince you? If not, please state what to do with the per-timeline eviction task logging. I'll implement that on top of `f480844^` then.

koivunej commented on libs/utils/src/rate_limit_closure.rs:
```diff
@@ -0,0 +1,66 @@
+
+    pub fn call<F: FnOnce()>(&mut self, f: F) {
+        let now = Instant::now();
+        match self.last {
+            Some(last) if now - last > self.interval => {
+                f();
+                self.last = Some(now);
+            }
+            None => {
+                f();
+                self.last = Some(now);
+            }
+            Some(_) => {}
+        }
```
otherwise if you really want the positive path only:

```suggestion
        let call_it = self.last.as_ref()
            .map(|x| now - last > self.interval)
            .unwrap_or(true);

        if call_it {
            self.last = Some(now);
            f();
        }
```

koivunej commented on pageserver/src/tenant/storage_layer.rs:
```diff
@@ -255,24 +300,44 @@ impl LayerAccessStats {
+    /// The `fallback` must be supplied for the case where there has not yet been a call to the
+    /// [`record_residence_event`] method.
+    /// If `fallback` needs to be used, we log a rate-limited warning in global scope.
+    pub(crate) fn latest_activity<F>(&self, fallback: F) -> SystemTime
```
```suggestion
    pub(crate) fn latest_activity<F>(&self) -> Option<SystemTime>
```

Would rather see this instead of new `Option::unwrap_or_else` alike APIs unless we use the value created by the factory for something within this function?

koivunej reviewed (changes requested):
I still dislike the approach (`RateLimitClosure` or after rename), static mutexes, but perhaps it truly is unavoidable or at least I don't have the energy to come up with alternative. At least the current review comments need to be addressed.

problame commented:
Updated the PR description to the final commit message.
